analyzers/src/SonarAnalyzer.Common/Rules/DeclareTypesInNamespacesBase.cs

```diff
@@ -18,34 +18,45 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
-using System;
+using System.Collections.Immutable;
 using Microsoft.CodeAnalysis;
-using Microsoft.CodeAnalysis.Diagnostics;
 using SonarAnalyzer.Helpers;
 
 namespace SonarAnalyzer.Rules
 {
-    public abstract class DeclareTypesInNamespacesBase : SonarDiagnosticAnalyzer
+    public abstract class DeclareTypesInNamespacesBase<TSyntaxKind> : SonarDiagnosticAnalyzer
+        where TSyntaxKind : struct
     {
         protected const string DiagnosticId = "S3903";
-        protected const string MessageFormat = "Move '{0}' into a named namespace.";
+        private const string MessageFormat = "Move '{0}' into a named namespace.";
 
-        protected abstract bool IsInnerTypeOrWithinNamespace(SyntaxNode declaration, SemanticModel semanticModel);
+        private readonly DiagnosticDescriptor rule;
+
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+        protected abstract TSyntaxKind[] SyntaxKinds { get; }
 
+        protected abstract bool IsInnerTypeOrWithinNamespace(SyntaxNode declaration, SemanticModel semanticModel);
         protected abstract SyntaxToken GetTypeIdentifier(SyntaxNode declaration);
 
-        protected Action<SyntaxNodeAnalysisContext> GetAnalysisAction(DiagnosticDescriptor rule) =>
-            c =>
-            {
-                var declaration = c.Node;
-                if (IsInnerTypeOrWithinNamespace(declaration, c.SemanticModel))
-                {
-                    return;
-                }
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected DeclareTypesInNamespacesBase() =>
+            rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, Language.RspecResources);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+          context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer,
+              c =>
+              {
+                  var declaration = c.Node;
 
-                var identifier = GetTypeIdentifier(declaration);
+                  if (IsInnerTypeOrWithinNamespace(declaration, c.SemanticModel) || c.ContainingSymbol.Kind != SymbolKind.NamedType)
```

Comment:
I think it's faster to do the symbol kind check first since we already have the symbol.